Add F# support

srivatsn opened the pull request:
This PR sets up the skeleton for what's needed to add basic F# support.

This change adds a new project system with a new project type guid. A IVsProjectSelector needs to be implemented to select this guid for a given project based on some condition of the project. To test though, the .sln file can be modified with this new guid and the project will open with the new project system.

Designtime targets: There's a new file called Microsoft.FSharp.DesignTime.targets that needs to be pulled in by the FSharp targets similar to how it's done for C# [here](https://github.com/Microsoft/msbuild/blob/xplat/src/Tasks/Microsoft.CSharp.CurrentVersion.targets#L325). This is so that the right CPS rules and capabilities get added to the project.

F# capability: The Designtime.targets adds a new FSharp capability. However all of the project system components today are marked as applicable for CSharp or VisualBasic. We need to go through them all and mark them as applicable to F# as well where it makes sense.

Language-service-hookup. There's three pieces of work here:
- The way the LS hookup works in roslyn project system is that kicks off a designtime build of a target called [CompileDesignTime](https://github.com/dotnet/roslyn-project-system/compare/master...srivatsn:fsharp?expand=1#diff-87271b4777d84b19f2d66ace1b99ac28R39) and expects to get the commandline string that would have been passed to the compiler. To do that the Fsc task needs to return the FscCommandLineArgs similar to how the csc task does that [here](https://github.com/dotnet/roslyn/blob/e509b57971474c4d5937bb9a7434839ee8e2c31b/src/Compilers/Core/MSBuildTask/Microsoft.CSharp.Core.targets#L142)
- Once the commandline is obtained, it needs to be parsed and FSharpCommandLineParserProvider needs to be filled out by calling into the F# compiler.
- Finally, we need to create a ProjectContext. There's common code that does that today. However it looks like Roslyn has a check for C#\VB in that code path and a different overload was added for F#. We should just fix that and standardize on calling with an error prefix.

Then there's the big workitem to show the items ordered by fileorder in the tree and provide gestures to move them around. 

I believe @brettfo @KevinRansom are going to fill out some of the pieces here.

@dotnet/project-system @Pilchie

srivatsn commented:
![fs](https://cloud.githubusercontent.com/assets/20570/23533152/a7b38640-ff64-11e6-9d76-06cc7160ddf2.jpg)

forki commented:
Kiss

forki commented:
I'm really excited for this. Thanks for doing this. 

One question after seeing the screenshot: will the package references work like in C#? I mean will it be possible to fill these in memory? This is important in order to keep paket dependency manager support. @davkean I assume yes, but would be nice to hear that everything will be the same.
Thx

davkean commented:
Package references will work exactly like C#.

forki commented:
Awesome thanks for confirmation.

enricosada commented:
> To do that the Fsc task needs to return the FscCommandLineArgs similar to how the csc task does that here

Already done that with https://github.com/dotnet/netcorecli-fsc/pull/88 (FSharp.NET.Sdk, the target file) and https://github.com/Microsoft/visualfsharp/pull/2523 (Fsc task)

@srivatsn if you want, you can use test it with the fsproj and nuget.config in https://github.com/enricosada/dotnet-proj-info/tree/master/examples as example. 
The nuget.config contains the dev feed with new fsctask and fsharp.net.sdk
the fsproj is the new tempalte for 1.0.0 rtm, so without dotnet-compile-fsc

i use that for a dotnetcli tool who will get fsc args `dotnet proj-info -p ..\c1\c1.fsproj --fsc-args`, ref https://github.com/enricosada/dotnet-proj-info/tree/master/examples

srivatsn commented:
@enricosada is there a plan to add that back to the regular desktop Fsc task as well? This project system should eventually work for all F# projects.

enricosada commented:
@srivatsn the package Fsharp.Net.Sdk version 1.0.0 already use only Fsc task (not dotnet compile fsc). Is the one from vf# repo. 

And also the new fsproj already works to create lib/console target framework net451 etc 

Works on mono too :)

enricosada commented:
And net45 are under integration test suite in https://github.com/dotnet/netcorecli-fsc/blob/master/test/dotnet-new.Tests/CommonScenarioTests.cs#L55


@srivatsn or you meant something different?

srivatsn commented:
Sorry I should have been clearer - the fact that the FscTask supports this now is great. However, I was referring to https://github.com/dotnet/netcorecli-fsc/pull/88. My understanding is that that PR is updating the FSharp.targets that are used in the FSharp.NET.Sdk which means any project using the SDK attribute and pulling that SDK will get this. However, old projects that exist today (created using previous versions of VS targeting desktop) will not create the commandline args. I think we should replace the existing project system with this effort and get all F# projects to open up with this new project system which would mean fixing up the targets in the [VF# repo](https://github.com/Microsoft/visualfsharp/blob/2d258b1209e88fedf29b46e6f0da6297de1ca4be/src/fsharp/FSharp.Build/Microsoft.FSharp.targets) - correct? So I was asking if you plan to PR those changes into the VF# repo.

To put this differently, I see that the F# SDK enables a bunch more things and i'm wondering if we should consolidate that with the targets in the VF# repo so that we don't have two styles of projects. For eg, one of the other workitems I mention above - adding an import the FSharp.DesignTime.targets that I'm adding in this PR needs to be done in the F# SDK and in the F# targets in the VF# repo before importing common.targets.

enricosada commented:
An update:

`FSharp.Compiler.Tools` 4.1.1 (used by `FSharp.NET.Sdk` 1.0.2) add the new properties to embedded `Fsc` task.
Eveything is on nuget.org, and the `dotnet new` templates (of .net core tools 1.0.1) resolve that now.

Ready to use examples also in https://github.com/enricosada/DotnetNewFsprojTestingSamples to test p2p and packageref

srivatsn commented:
Thanks @enricosada - I just tried the FSharp.NET.Sdk 1.0.2. There's still a problem. It looks like the new properties have been added to the netcoreapp1.0 version of the Fsc task in FSharp.Build.dll but not to the desktop version. Since VS uses desktop msbuild, it picks up the Fsharp.Build.dll in the tools folder in that package and fails design time builds fail saying it can't find the SkipCompilerExecution property.

srivatsn commented:
@dotnet/project-system I think this is ready for review. I've cleaned up some of the hacks, added tests and created issues for the remaining work - 

#1853 - F# language service integration 
#1854 - F# property pages 
#1855 - F# item templates don't show up 
#1856 - F# resx and xaml files need to be localized and setup authored. 
#1857 - F# SDK needs to import design-time targets and authored into VS

saul commented on src/Targets/Microsoft.FSharp.DesignTime.targets:
```diff
@@ -0,0 +1,48 @@
+
+  <!-- Targets -->
+
+  <!-- Returns Csc command-line arguments for the language service -->
```
Csc -> Fsc

saul commented:
@srivatsn this looks great, thanks for doing this :)

Out of interest, what needs to be completed from the F# front? i.e. is there any support for arbitrary ordering in the solution explorer? I'm sure this has been mentioned before but I can't remember where I saw it.

enricosada commented:
> I just tried the FSharp.NET.Sdk 1.0.2. There's still a problem. It looks like the new properties have been added to the netcoreapp1.0 version of the Fsc task in FSharp.Build.dll but not to the desktop version. Since VS uses desktop msbuild, it picks up the Fsharp.Build.dll in the tools folder in that package and fails design time builds fail saying it can't find the SkipCompilerExecution property

yes, is an issue also for mono. Is tracked in https://github.com/fsharp/fsharp/issues/684

srivatsn commented:
> is there any support for arbitrary ordering in the solution explorer?
No, that needs to be implemented as well. Filed https://github.com/dotnet/roslyn-project-system/issues/1875 to track that.

saul commented:
Thanks @srivatsn. 

What's the easy way to get setup and working on these tickets? I'm happy to do some of the work, especially as I just spent a few days working on the file ordering in the old project system.

What's your dev cycle at the minute?

cc @brettfo - are these tickets open to be worked on by the community at this point, or just MS?

srivatsn commented:
@saul - sure feel free to pick up any of this work. Let @brettfo know since I believe he plans to tackle them otherwise. https://github.com/dotnet/roslyn-project-system/blob/master/docs/repo/getting-started.md has instructions to get started.

brettfo commented:
Also to @saul (and @srivatsn), one of the steps will be to provide an F# command line parser.  I've created #1876 which starts this work, but that's ultimately dependent upon some changes in Roslyn.  Once both of these are merged, I'll start work on an F# command line parser that can be used here.

cloudRoutine commented:
@brettfo like this one?
https://github.com/cloudRoutine/netcorecli-fsc/blob/fsharp-impl/src/dotnet-compile-fsc/Program.fs
https://github.com/cloudRoutine/netcorecli-fsc/tree/fsharp-impl/src/dotnet-compile-fsc

or does it need to use a different API?

brettfo commented:
@saul, Speaking with @KevinRansom he suggested factoring out [this](https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/fsc.fs#L1635) method.  My plan was to add a type `FSharpCommandLineArguments that inherts from the Roslyn type `CommandLineArguments`, but first some APIs need to be opened up (see here: dotnet/roslyn#18258).

Once that's been done I need to consume the F# command line parser [here](https://github.com/dotnet/roslyn-project-system/pull/1876/commits/d82b0087f7bd263e9eb6ba7eff5f1a04c03cd77c).

saul commented:
@brettfo @srivatsn do you have an example project file?

I've followed the instructions (using RoslynDev suffix) and I'm using one of @enricosada's example F# .Net Core projects, but I'm getting the following error:

![image](https://cloud.githubusercontent.com/assets/64654/24426995/5c3da0fe-1401-11e7-80af-c830823cceda.png)

srivatsn commented:
@saul Until https://github.com/dotnet/roslyn-project-system/issues/1857 is fixed, you'll need to make those changes in your machine. That is, you need to:
- Copy the FSharp.NET.Sdk folder from %Program Files%\Dotnet\SDK\sdks to %vsdir%\msbuild\Sdks
- Add the snippet of xml mentioned in #1857 to the two FSharp.NET.Sdk (one in the VSDIR and the other one in your nuget packages directory).
- Then you can open a F# netcore project by either renaming to msbuildproj as you seem to be doing or by changing the project type guid in the .sln file from `F2A71F9B-5D33-465A-A702-920D77279786` (Existing project system's guid) to `6EC3EE1D-3C4E-46DD-8F32-0CC8E7565705` (new project system's guid)

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/Packaging/FSharpProjectSystemPackage.cs:
```diff
@@ -0,0 +1,44 @@
+{
+    [Guid(PackageGuid)]
+    [PackageRegistration(AllowsBackgroundLoading = true, RegisterUsing = RegistrationMethod.CodeBase, UseManagedResourcesOnly = true)]
+    [RemoteCodeGeneratorRegistration(SingleFileGenerators.ResXGuid, SingleFileGenerators.ResXGeneratorName,
```
It looks like you got these off C#, are these the same ones that are enabled for existing F# proejct system?

srivatsn replied:
You are right - the old F# PS doesn't seem to register any singlefilegenerators. These likely produce cs files and are incorrect. I've removed them.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/Properties/FSharpProjectDesignerPage.cs:
```diff
@@ -0,0 +1,19 @@
+    internal static class FSharpProjectDesignerPage
+    {
+        public static readonly ProjectDesignerPageMetadata Application = new ProjectDesignerPageMetadata(new Guid("{6D2D9B56-2691-4624-A1BF-D07A14594748}"), pageOrder: 0, hasConfigurationCondition: false);
+        public static readonly ProjectDesignerPageMetadata Build = new ProjectDesignerPageMetadata(new Guid("{FAC0A17E-2E70-4211-916A-0D34FB708BFF}"), pageOrder: 1, hasConfigurationCondition: false);
```
unify these?

srivatsn replied:
Unify the pages you mean? I think the build page is different for F#.

xperiandri replied:
Because of this?
![image](https://cloud.githubusercontent.com/assets/2365592/24443693/4dc5a128-146c-11e7-97dd-47c0065b6ab9.png)

xperiandri replied:
Is it XAML? Could you make this page shrinking horizontally?
Textboxes can become wider without any problem so that all the content can be visible.
This with limitation is very inconvenient!

srivatsn replied:
Ha - is that the only difference? I just meant that the implementation is different (a separate class with a separate guid in the F# assembly but didn't realize it's that close to C#). Ya it might be nice to consolidate these.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/VsContainedLanguageComponentsFactory.cs:
```diff
@@ -14,7 +14,7 @@
 {
     [Export(typeof(IVsContainedLanguageComponentsFactory))]
-    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasicOrFSharp)]
```
I reckon we should mark these as "Managed".

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/Microsoft.VisualStudio.ProjectSystem.Managed.csproj:
```diff
@@ -33,6 +37,9 @@
     <Compile Update="ProjectSystem\Rules\AdditionalFiles.cs">
       <DependentUpon>AdditionalFiles.xaml</DependentUpon>
     </Compile>
+    <Compile Update="ProjectSystem\Rules\FSharp.cs">
```
We get a bug to pull these into appropriate language-specific projects.

davkean commented on src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/FSharp.BrowseObject.xaml:
```diff
@@ -0,0 +1,99 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<!--Copyright, Microsoft Corporation, All rights reserved.-->
```
Should we just unify these between C#/VB and F#? Do they have any different properties?

davkean replied:
Actually, why can we just use the C# one? Don't we do that for VB? And just rename the thing to "Compile"

srivatsn replied:
Will do as a separate PR since it touches too many files.

enricosada commented:
as a note:

- https://github.com/dotnet/netcorecli-fsc/pull/101 to change the guid in f# to use new (i will release this and others needed stuff with `FSharp.NET.Sdk` v`1.0.3`)
- https://github.com/dotnet/netcorecli-fsc/pull/99 will use fsc.exe (so .net full) when msbuild host is .net full (msbuild.exe), like in VS.
- https://github.com/dotnet/netcorecli-fsc/pull/100 update `FSharp.Build` tasks with new propeties, also for .NET full fsc tasks

enricosada commented:
@srivatsn there are `<Capability` i need to add to `FSharp.NET.Sdk`? if i understand right how that works..

basoundr commented on src/Microsoft.VisualStudio.ProjectSystem.FSharp.VS/ProjectSystem/VS/FSharpProjectGuidProvider.cs:
```diff
@@ -0,0 +1,30 @@
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    /// <summary>
+    ///     Provides the Visual Basic implementation of <see cref="IItemTypeGuidProvider"/>.
```
`Visual Basic` -> `FSharp`

basoundr replied:
I guess there are lot of places where Visual Basic or C# will be present instead of F#. As we implement the component we should update these files.

srivatsn commented:
@enricosada the capabilities are included in the FSharp.DesignTime.targets that I've added in this PR. This targets file needs to be imported by the FSharp SDK. More details in https://github.com/dotnet/roslyn-project-system/issues/1857
